functions/src/storeTrackingData.ts

```diff
@@ -1,4 +1,100 @@
+export {};
+const { google } = require('googleapis');
+const functions = require('firebase-functions');
+const GenderApi = require('gender-api.com-client');
+
+function getAppDate(appDate: any) {
+  const monthNames = [
+    'Jan',
+    'Feb',
+    'Mar',
+    'Apr',
+    'May',
+    'Jun',
+    'Jul',
+    'Aug',
+    'Sep',
+    'Oct',
+    'Nov',
+    'Dec',
+  ];
+  const newdate =
+    appDate.getDate() +
+    '-' +
+    monthNames[appDate.getMonth()] +
+    '-' +
+    appDate.getFullYear();
+  return newdate;
+}
+
 exports.handler = async function(snap: any) {
   const email: string = snap.data()!.email;
-  console.log(email);
+  const name: string = snap.data()!.name;
+  const batch: string = snap.data()!.batch;
+  const userType: string = snap.data()!.userType;
+  const track: string = snap.data()!.track;
+  const source: string = snap.data()!.source;
+  const applicationTime: any = getAppDate(snap.createTime.toDate());
+
+  const OAuth2 = google.auth.OAuth2;
+  const clientID = functions.config().spreadsheet.clientid;
+  const clientSecret = functions.config().spreadsheet.clientsecret;
+  const refreshToken = functions.config().spreadsheet.refreshtoken;
+
+  const oauth2Client = new OAuth2(
+    clientID, //client Id
+    clientSecret, // Client Secret
+    'https://developers.google.com/oauthplayground' // Redirect URL
+  );
+
+  oauth2Client.setCredentials({
+    refresh_token: refreshToken,
+  });
+
+  function appendPromise(requestWithoutAuth: any) {
+    console.log('appendpromisee');
+    return new Promise((resolve, reject) => {
+      const sheets = google.sheets('v4');
+      const request = requestWithoutAuth;
+      return sheets.spreadsheets.values.append(
+        request,
+        (err: any, response: any) => {
+          if (err) {
+            console.log(`The API returned an error: ${err}`);
+            reject(err);
+            return;
+          }
+          resolve(response.data);
+          return;
+        }
+      );
+    });
+  }
+
+  // api used: https://gender-api.com/en/clients
+  const genderApiClient = new GenderApi.Client('VYdUEnwvvPsSKUeFwD');
+  genderApiClient.getByFirstName(name, function(response: any) {
+    console.log(response.gender);
+    void appendPromise({
```

Comment:
I think the function name here is misleading, we could call the sheets API directly in a callback manner, no need to wrap in promise as far as I see, because we don't take advantage of the fact that it's wrapped in promise

Reply:
So calling it like this, without the function, should be fine as well

```js
sheets.spreadsheets.values.append(
  request,
  (err: any, response: any) => {
    if (err) {
      console.log(`The API returned an error: ${err}`);
      reject(err);
      return;
    }
    resolve(response.data);
    return;
  }
```

Reply:
- So I just pushed the requested changes. I hope this is what you meant. 
- Regarding the API used: I want to know do we plan to use this gender api or not? Because if yes then we should get a proper account running. It gives every month free 500 API requests quota but I guess we need more maybe? 
- If you think everything works fine, then I'd store the correct google sheet url and new API key in a config file.